core/js/src/typespecs/prompt.ts

```diff
@@ -0,0 +1,122 @@
+import { z } from "zod";
+
+export const roleSchema = z.enum([
+  "system",
+  "user",
+  "assistant",
+  "function",
+  "tool",
+  "model",
+]);
+export type Role = z.infer<typeof roleSchema>;
+
+export const functionCallSchema = z.object({
+  name: z.string(),
+  arguments: z.string(),
+});
+
+const toolCallSchema = z.object({
+  id: z.string(),
+  function: z.object({
+    arguments: z.string(),
+    name: z.string(),
+  }),
+  type: z.literal("function"),
+});
+
+export const messageSchema = z.object({
+  content: z.string().default(""),
+  role: roleSchema,
+  name: z.string().optional(),
```

Comment:
Usually I prefer `nullish` to `optional`, because it lets people pass in `null` as the value. In fact a lot of auto-generated SDKs (such as python) will explicitly pass in `None`, which gets converted to `null`, and then rejected by an API spec that only allows for `optional`.